pkg/ast/spl/spl.peg

```diff
@@ -376,6 +376,168 @@ SearchTerm <- term:(FieldWithNumberValue / FieldWithStringValue) {
+            UseOther: false,
+            OtherStr: "Other",
+        }
+    }
+
+    if byClause != nil {
+        statisticExpr.ByClause = byClause.([]string)
+    }
+
+    return statisticExpr, nil
+}
+
+//Top limit=<int> is the same as specifying top N.
+StatisticLimit <- SPACE number:IntegerAsString {
+    return number.(string), nil
+}
+/ limit:(SPACE "limit=" IntegerAsString) {
+    return limit.([]interface{})[2].(string), nil
+}
+
+//Optional arguments for top/rare functions
+Options <- option:(Option)*
+{
+    //Default value
+    options := &structs.Options {
+        ShowCount: true,
+        CountField: "count",
+        ShowPerc: true,
+        PercentField: "percent",
+        UseOther: false,
+        OtherStr: "other",
+    }
+
+    optionSlice := option.([]any)
+    for _, opt := range optionSlice {
+        optArr :=  opt.([]string)
+        // 使用类型断言检查匹配的选项类型
+        switch optArr[0] {
+        case "showcount":
+            showCountBool, err := strconv.ParseBool(optArr[1])
+            if err != nil {
+                return nil, fmt.Errorf("Spl peg: StatisticBlock: Options: %v", err)
+            } 
+            options.ShowCount = showCountBool
+        case "countfield":
+            options.CountField = optArr[1]
+        case "showperc":
+            showPercBool, err := strconv.ParseBool(optArr[1])
+            if err != nil {
+                return nil, fmt.Errorf("Spl peg: StatisticBlock: Options: %v", err)
+            } 
+            options.ShowPerc = showPercBool
+        case "percentfield":
+            options.PercentField = optArr[1]
+        case "useother":
+            useOtherBool, err := strconv.ParseBool(optArr[1])
+            if err != nil {
+                return nil, fmt.Errorf("Spl peg: StatisticBlock: Options: %v", err)
+            } 
+            options.UseOther = useOtherBool
+        case "otherstr":
+            options.OtherStr = optArr[1]
+        default:
+            return nil, fmt.Errorf("Spl peg: StatisticBlock: Options: invalid option")
+        }
+    }
+
+    return options, nil
+}
+
+Option <- SPACE optionCMD:OptionCMD "=" field:EvalFieldToRead {
+    optionArr := []string{optionCMD.(string), field.(string)}
+    return optionArr, nil
+}
+
+OptionCMD <- option:("countfield" / "showcount" / "otherstr" / "useother"/ "percentfield" / "showperc") {
+    optionStr, err := transferUint8ToString(option)
+    if err != nil {
+        return nil, fmt.Errorf("Spl peg: StatisticExpr: %v", err)
+    }
+    return optionStr, nil
+}
+
+FieldList <- SPACE field:EvalFieldToRead rest:("," SPACE EvalFieldToRead)* {
+    restSlice := rest.([]any)
+    slice := make([]string, 1 + len(restSlice))
+    slice[0] = field.(string)
+
+    for i := range restSlice {
+        fieldAtom := restSlice[i].([]any)
+        slice[i + 1] = fieldAtom[2].(string)
+    }
+
+    return slice, nil
+}
+
+ByClause <- SPACE "by" fieldList:FieldList {
+    return fieldList.([]string), nil
+}
```

Comment:
Can we use the existing `GroupbyBlock` rule instead, or is that rule different enough that we have to make these rules?

Reply:
Yes, it is almost the same. In the Splunk doc, keyword "by" in example is in lower case. If it does not matter, I can use GroupBlock instead.

Reply:
In the Splunk documentation here: https://docs.splunk.com/Documentation/Splunk/9.1.1/SearchReference/Top
They have this:
> <by-clause\>
Syntax: BY <field-list>
Description: The name of one or more fields to group by.

But in the examples they use lowercase `by`. So we should handle both upper and lower case. For the `GroupbyBlock` I think lowercase should also be allowed but I'm not quite sure. So you can change `GroupbyBlock` to allow upper and lower case, and then reuse that rule here.